### PR TITLE
#576 Unique Id Mapper should not return an option

### PR DESCRIPTION
**Pull request by anilgursel**
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [ ] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala [hunk 1]
```diff
@@ -71,7 +71,7 @@ object RetryBidi {
     * @tparam Context the type of the context that is carried along with the elements.
     * @return a [[BidiFlow]] with Retry functionality
     */
-  def apply[In, Out, Context](maxRetries: Int, uniqueIdMapper: Context => Option[Any] = (_: Any) => None,
+  def apply[In, Out, Context](maxRetries: Int, uniqueIdMapper: Context => Any = (c: Context) => c,
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala]: I think it is more clear and compact if you say `uniqueIdMapper: Context => Any = identity`. Please try it out.

File: squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala [hunk 2]
```diff
@@ -300,11 +300,9 @@ final class TimeoutBidiUnordered[In, Out, Context](timeout: FiniteDuration,
...
 
   private[streams] def uniqueId(context: Context) =
-    uniqueIdMapper(context).getOrElse {
-      context match {
-        case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
-        case context ⇒ context
-      }
+    uniqueIdMapper(context) match {
+      case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
+      case uniqueId ⇒ uniqueId
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala]: One reason for getting rid of `Context => Option[Any]` is to optimize the decision to be just once. I don't think we have achieved this just yet. If you would model it as `Option[Context => Any]` then you could make this decision upfront. Does not mean the API needs to show `Option` or `Optional`. That we can do behind the scenes instead of a function that gets called all the time. Just for consideration...

File: squbs-ext/src/main/scala/org/squbs/streams/UniqueId.scala [hunk 3]
```diff
@@ -37,11 +37,8 @@ object UniqueId {
...
   }
 
-  private[squbs] def javaUniqueIdMapperAsScala[Context](uniqueIdMapper: JFunction[Context, Optional[Any]]):
-  Context => Option[Any] = {
-
-    import scala.compat.java8.OptionConverters._
-
-    (context: Context) => toScala(uniqueIdMapper(context))
+  private[squbs] def javaUniqueIdMapperAsScala[Context](uniqueIdMapper: JFunction[Context, Any]):
+  Context => Any = {
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/UniqueId.scala]: `... Context => Any = uniqueIdMapper.apply`

May need an `_` after the apply just in case the compiler does not recognize it.

File: squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala [hunk 4]
```diff
@@ -60,7 +60,7 @@ case class CircuitBreakerSettings[In, Out, Context] private[japi] (
       fallback.asScala.map(f => (in: In) => f(in)),
       (out: Out) => cleanUp.accept(out),
       failureDecider.asScala.map(f => (out: Try[Out]) => f(out).asInstanceOf[Boolean]),
-      UniqueId.javaUniqueIdMapperAsScala(uniqueIdMapper))
+      uniqueIdMapper.asScala)
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala]: Ditto. You don't need a function converter to convert a Java 8 `Function` to Scala. Just the `Function.apply _` is by itself a Scala function.

File: squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala (hunk 2 above)
Comment (anilgursel) [reply squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala]: The only thing we could save is the `match`, we still need to call the function.  Because the `match` won't change based on the request.

However, do we want to optimize not calling `uniqueIdMapper` when one not provided?  I would assume if we set it to `identity`, compiler would do the optimization.

Comment (akara) [reply squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala]: Yes, I also assume `identity` will reach that same result. Compiler should optimize it away. So if we can make sure the match does not happen for every message, I'm good.

File: squbs-ext/src/main/scala/org/squbs/streams/UniqueId.scala (hunk 3 above)
Comment (anilgursel) [reply squbs-ext/src/main/scala/org/squbs/streams/UniqueId.scala]: I had actually eliminated all uses of this function, but apparently, forgot to delete it.

**Comment by anilgursel** [comment]
@akara Please check if the last commit addresses the concerns.  Please note, I am not very happy how the API currently looks and also, there is some inconsistency with `TimeoutBidiUnordered` (which takes an `Option`).   However, I will address the concerns with another PR as follows:

* Rename stages as `Retry`, `CircuitBreaker`, `Timeout`.
* Each takes a `Settings`, defined inside the corresponding companion object.
* Retry, CircuitBreaker and Timeout will have APIs that take settings, but also their simple form, e.g., Retry(5).

File: docs/flow-timeout.md [hunk 5]
```diff
@@ -126,7 +126,7 @@ The following API is used to pass a uniqueId mapper:
 
 ```scala
 TimeoutBidiFlowUnordered[In, Out, Context](timeout: FiniteDuration,
-                                           uniqueIdMapper: Context => Option[Any])
+                                           uniqueIdMapper: OPtion[Context => Any])
```
Comment (akara) [inline docs/flow-timeout.md]: Eh? Typo in `OPtion`?

File: squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala [hunk 6]
```diff
@@ -204,13 +163,12 @@ final class RetryBidi[In, Out, Context] private[streams](maxRetries: Int, unique
   require(exponentialBackoffFactor >= 0.0, "backoff factor must be >= 0.0")
   require(maxDelay == Duration.Zero || maxDelay >= delay, "maxDelay must be larger than delay")
 
-  private[streams] def uniqueId(context: Context) =
-    uniqueIdMapper(context).getOrElse {
-      context match {
-        case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
-        case `context` ⇒ `context`
-      }
+  val uniqueId: Context => Any = uniqueIdMapper.getOrElse{
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala]: Yes, I do like this much more. Thanks!

File: squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala [hunk 7]
```diff
@@ -299,13 +298,12 @@ final class TimeoutBidiUnordered[In, Out, Context](timeout: FiniteDuration,
   private val out = Outlet[(Try[Out], Context)]("TimeoutBidiUnordered.out")
   val shape = BidiShape(in, toWrapped, fromWrapped, out)
 
-  private[streams] def uniqueId(context: Context) =
-    uniqueIdMapper(context).getOrElse {
-      context match {
-        case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
-        case context ⇒ context
-      }
+  val uniqueId: Context => Any = uniqueIdMapper.getOrElse{
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala]: 👍

File: squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala [hunk 8]
```diff
@@ -423,14 +381,14 @@ final class RetryBidi[In, Out, Context] private[streams](maxRetries: Int, unique
   */
 case class RetrySettings[In, Out, Context] private[streams](
    maxRetries: Int,
-   uniqueIdMapper: Context => Option[Any] = (_: Any) => None,
+   uniqueIdMapper: Option[Context => Any] = None,
```
Comment (akara) [inline squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala]: We wanted to rename `RetrySettings` to `Retry.Settings`, right? Is that supposed to be in a different PR?

Comment (anilgursel) [reply squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala]: Yes, that will be a separate PR.
